tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/testing/HeadlessTester.java

```diff
@@ -0,0 +1,202 @@
+					if (report.exists()) {
+						try {
+							FileDeleter.delete(report);
+						} catch (IOException e) {
+							throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+									"Test report location cannot be cleared at: "
+											+ report.getAbsolutePath() + ".",
+									e);
+						}
+					}
+					if (testListener.finished()) {
+						TestReport testReport = new TestReport((StringBuilder) testTreeXmlTransformer.apply(testTree));
+						try {
+							testReport.dump(report);
+						} catch (IOException e) {
+							throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+									"Test report location cannot be generated at: "
+											+ report.getAbsolutePath() + ".",
+									e);
+						}
+						System.out.println("_____________");
+					} else {
+						// thread.sleep to wait for the test event busy?
+						errors += "test session still in progress.\n";
+					}
+				} else {
+					// thread.sleep to wait for the test event busy?
+					errors += "cannot write test report to " + testReportRoot + "\n";
+				}
+
+			}
+
+			if (!testListener.isOK()) {
+				errors += "There were test errors, see console logs and/or test report for details.\n";
+			}
+
+			if (!errors.isEmpty()) {
+				throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+						errors);
+			}
+
+		} catch (InterruptedException e1) {
+			logger.error(Throwables.getStackTraceAsString(e1));
+			throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+					"The spawned tester exited by throwing an exception", e1);
+		} finally {
+			testerFacade.shutdownFramework();
+		}
+	}
+
+	private ITesterDescriptor checkTester(String tester) throws ExitCodeException {
+		final List<ITesterDescriptor> matchingTesterDescs = findTesterById(tester);
+
+		if (matchingTesterDescs.isEmpty()) {
+			throw new ExitCodeException(EXITCODE_TESTER_NOT_FOUND, "no tester found for id: " + tester);
+		} else if (matchingTesterDescs.size() > 1) {
+			throw new ExitCodeException(EXITCODE_TESTER_NOT_FOUND, "several testers match given id \"" + tester
+					+ "\":\n\t" + Joiner.on("\n\t").join(matchingTesterDescs));
+		}
+
+		return matchingTesterDescs.get(0);
+	}
+
+	private List<ITesterDescriptor> findTesterById(String testerId) {
+		if (testerId == null || testerId.trim().isEmpty())
+			return Collections.emptyList();
+		// 1st attempt: look for exact match
+		final ITesterDescriptor td = testerRegistry.getDescriptors().get(testerId);
+		if (td != null)
+			return Collections.singletonList(td);
+
+		logger.warn("Could not find tester by ID: " + testerId + ", switching to fuzzy search.");
+
+		// 2nd attempt: look for sloppy match (but full segments required!)
+		final int t_len = testerId.length();
+		final List<ITesterDescriptor> matchingTDs = testerRegistry.getDescriptors().values().stream()
+				.map(descriptor -> descriptor.getId())
+				.filter(id -> {
+					final int id_len = id.length();
+					return id_len >= t_len
```

Comment:
Understandability: first: `.getDescriptors().values()`, then mapping, then filtering. The filtering uses an anonymous function that has no name and tries to be understandable by using some comments. Later, also `peek`, `map` again and `collect` are called. All in all, this statement spans across the lines 186-198. I think, this should be improved.